### PR TITLE
AoC 2019 Day 9 "IntCode generalisation, double-tapped the puzzle"

### PR DESCRIPTION
**Pull request by P-on-D**
Loved this one.

### CODE REVIEW

File: 2019/intcode.d [hunk 1]
```diff
@@ -1,66 +1,90 @@
...
 
-    final switch (instr.op) {
-      case Opcode.JT:  ld(); PC = (instr.ld1 != 0) ? instr.ld2 : (PC + 3); return true;
-      case Opcode.JF:  ld(); PC = (instr.ld1 == 0) ? instr.ld2 : (PC + 3); return true;
-      case Opcode.Add: ld(); memory[memory[PC+3]] = instr.ld1 + instr.ld2; PC = PC + 4; return true;
-      case Opcode.Mul: ld(); memory[memory[PC+3]] = instr.ld1 * instr.ld2; PC = PC + 4; return true;
-      case Opcode.Lt:  ld(); memory[memory[PC+3]] = instr.ld1 < instr.ld2; PC = PC + 4; return true;
-      case Opcode.Eq:  ld(); memory[memory[PC+3]] = instr.ld1 == instr.ld2; PC = PC + 4; return true;
+    final switch (op) {
+      case Opcode.JT:  PC = (ld(1, m1) != 0) ? ld(2, m2) : (PC + 3); return true;
```
Comment (P-on-D-Reviewer) [inline 2019/intcode.d:65]: Although this looks clean, the [compiler](https://gcc.godbolt.org/z/RwGPfh) is generating actual function calls here which can't be helping performance. The modal memory access code is duplicated. Is there a way to ask the runtime for a memory allocator that takes care of this for you?

File: 2019/intcode.d [hunk 2]
```diff
@@ -1,66 +1,90 @@
...
 
-struct IntCode {
-  int[] memory;
-  int[] input;
-  int[] output;
+alias IntCode = IntCodeT!int;
+alias LongCode = IntCodeT!long;
 
-  alias Addr = int;
+struct IntCodeT(T) {
```
Comment (P-on-D-Reviewer) [inline 2019/intcode.d:6]: This should have a formal constructor now that copies the program.

**Review by P-on-D-Reviewer — CHANGES_REQUESTED** [review]
Looks good, some changes to consider

File: 2019/intcode.d (hunk 1 above)
Comment (P-on-D) [reply 2019/intcode.d:65]: `std.container.array` looks promising but would expose an interface change (`IntCode.memory` is no longer an array).

**Review by P-on-D-Reviewer — APPROVED** [review]

**Pull request merged**